events/ratelimiting/coalescing.go

```diff
@@ -0,0 +1,256 @@
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"k8s.io/utils/clock"
+)
+
+// OptionsCoalescing configures a Coalescing RateLimiter.
+type OptionsCoalescing struct {
+	// InitialDelay is the initial delay for the rate limiter. The rate limiter
+	// will not delay events less than the initial delay.
+	// Defaults to 500ms.
+	InitialDelay *time.Duration
+
+	// MaxDelay is the maximum delay for the rate limiter. The rate limiter will
+	// not delay events longer than the max delay.
+	// Defaults to 5s.
+	MaxDelay *time.Duration
+
+	// MaxPendingEvents is the maximum number of events that can pending on a
+	// rate limiter, before it fires an event anyway. Useful to prevent a rate
+	// limiter never firing events in a high throughput scenario.
+	// Defaults to unlimited.
+	MaxPendingEvents *int
+}
+
+// coalescing is a rate limiter that rate limits events. It coalesces events
+// that occur within a rate limiting window.
+type coalescing struct {
+	initialDelay     time.Duration
+	maxDelay         time.Duration
+	maxPendingEvents *int
+
+	pendingEvents int
+	timer         clock.Timer
+	hasTimer      atomic.Bool
+	inputCh       chan struct{}
+	currentDur    time.Duration
+	backoffFactor int
+
+	wg      sync.WaitGroup
+	lock    sync.RWMutex
+	clock   clock.WithTicker
+	running atomic.Bool
+	closeCh chan struct{}
+	closed  atomic.Bool
+}
+
+func NewCoalescing(opts OptionsCoalescing) (RateLimiter, error) {
+	initialDelay := time.Millisecond * 500
+	if opts.InitialDelay != nil {
+		initialDelay = *opts.InitialDelay
+	}
+	if initialDelay < 0 {
+		return nil, errors.New("initial delay must be > 0")
+	}
+
+	maxDelay := time.Second * 5
+	if opts.MaxDelay != nil {
+		maxDelay = *opts.MaxDelay
+	}
+	if maxDelay < 0 {
+		return nil, errors.New("max delay must be > 0")
+	}
+
+	if maxDelay < initialDelay {
+		return nil, errors.New("max delay must be >= base delay")
+	}
+
+	if opts.MaxPendingEvents != nil && *opts.MaxPendingEvents <= 0 {
+		return nil, errors.New("max pending events must be > 0")
+	}
+
+	return &coalescing{
+		initialDelay:     initialDelay,
+		maxDelay:         maxDelay,
+		maxPendingEvents: opts.MaxPendingEvents,
+		currentDur:       initialDelay,
+		backoffFactor:    1,
+		inputCh:          make(chan struct{}),
+		closeCh:          make(chan struct{}),
+		clock:            clock.RealClock{},
+	}, nil
+}
+
+// Run runs the rate limiter. It will begin rate limiting events after the
+// first event is received.
+func (c *coalescing) Run(ctx context.Context, ch chan<- struct{}) error {
+	if !c.running.CompareAndSwap(false, true) {
+		return errors.New("already running")
+	}
+
+	// Prevent wg race condition on Close and Run.
+	c.lock.Lock()
+	c.wg.Add(1)
+	c.lock.Unlock()
+	defer c.wg.Done()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for {
+		// If the timer doesn't exist yet, we're waiting for the first event (which
+		// will fire immediately when received).
+		if !c.hasTimer.Load() {
```

Comment:
I think this could be simplified to keep it DRY (sing ereading from a nil channel is allowed and it never receives a message)

```go
        var timerCh <-chan Time
		if c.hasTimer.Load() {
			timerCh = c.timer.C()
		}

		select {
		case <-ctx.Done():
			return nil
		case <-c.closeCh:
			cancel()
			return nil

		case <-c.inputCh:
			c.handleInputCh(ctx, ch)
		case <-timerCh:
			c.handleTimerFired(ctx, ch)
		}
```

Reply:
Thanks, this is a lot cleaner :+1: